msbuild/Xamarin.MacDev.Tasks.Core/Tasks/CodesignTaskBase.cs

```diff
@@ -194,15 +241,107 @@ public override bool Execute ()
+
+			// Then we need to split the input into buckets, where everything in a bucket can be signed in parallel
+			// (i.e. no item in a bucket depends on any other item in the bucket being signed first).
+			// any such items must go into a different bucket. The bucket themselves are also sorted, where
+			// we have to sign the first bucket first, and so on.
+			// Since we've sorted by path length, we know that if we find a directory, we won't find any containing
+			// files from that directory later.
+			var buckets = new List<List<ITaskItem>> ();
+			for (var i = 0; i < resourcesToSign.Length; i++) {
+				var res = resourcesToSign [i];
+				// All files can go into the first bucket.
+				if (File.Exists (res.ItemSpec)) {
+					if (buckets.Count == 0)
+						buckets.Add (new List<ITaskItem> ());
+					var bucket = buckets [0];
+					bucket.Add (res);
+					continue;
+				}
+
+				if (Directory.Exists (res.ItemSpec)) {
+					var dir = res.ItemSpec;
+
+					// Add the directory separator, so we can do easy substring matches
+					if (dir.Length > 0 && dir [dir.Length - 1] != Path.DirectorySeparatorChar)
+						dir += Path.DirectorySeparatorChar;
+
+					// This is a directory, which can contain other files or directories that must be signed first
+					// If this item is a containing directory for any of the items in a bucket, then we need to
+					// add this item to the next bucket. So we go through the buckets in reverse order.
+					var added = false;
+					for (var b = buckets.Count - 1; b >= 0; b--) {
+						var bucket = buckets [b];
+						var anyContainingFile = bucket.Any (v => v.ItemSpec.StartsWith (dir, StringComparison.OrdinalIgnoreCase));
+						if (anyContainingFile) {
+							if (b + 1 >= buckets.Count)
+								buckets.Add (new List<ITaskItem> ());
+							buckets [b + 1].Add (res);
+							added = true;
+							break;
+						}
+					}
+					if (!added) {
+						// This directory doesn't contain any other signed files, so we can add it to the first bucket.
+						if (buckets.Count == 0)
+							buckets.Add (new List<ITaskItem> ());
+						var bucket = buckets [0];
+						bucket.Add (res);
+					}
+					continue;
+				}
 
-			Parallel.ForEach (resourcesToSign, new ParallelOptions { MaxDegreeOfParallelism = Math.Max (Environment.ProcessorCount / 2, 1) }, (item) => {
-				Codesign (item);
+				Log.LogWarning ("Unable to sign '{0}': file or directory not found.", res.ItemSpec);
+			}
 
-				var files = GetCodesignedFiles (item);
-				lock (codesignedFiles)
-					codesignedFiles.AddRange (files);
-			});
+#if false
+			Log.LogWarning ("Codesigning {0} buckets", buckets.Count);
+			for (var b = 0; b < buckets.Count; b++) {
+				var bucket = buckets [b];
+				Log.LogWarning ($"    Bucket #{b + 1} contains {bucket.Count} items:");
+				foreach (var item in bucket) {
+					Log.LogWarning ($"        {item.ItemSpec}");
+				}
+			}
+#endif
+
+			for (var b = 0; b < buckets.Count; b++) {
```

Comment:
Same, lets get index out, not really used elsewhere, right?

Reply:
Same reason: the order is important, and an enumerable doesn't have a guaranteed order.